File: services/notification/app/metrics.py
"""Prometheus metrics: RED for HTTP, plus the async-worker series.

The `route` label is the matched route TEMPLATE (e.g. "/notifications"), never the
raw path, so values don't explode metric cardinality. These series feed your SLO
and Grafana RED dashboards.

Per api-details §1.3, async workers also expose `events_consumed_total{topic,result}`
and `event_processing_duration_seconds{topic}`. We add `notifications_sent_total`
as this service's domain counter.
"""
from __future__ import annotations

from prometheus_client import Counter, Histogram

# --- RED (HTTP) ---
http_requests_total = Counter(
    "http_requests_total",
    "Total number of HTTP requests processed.",
    ["route", "method", "status"],
)

http_request_duration_seconds = Histogram(
    "http_request_duration_seconds",
    "HTTP request latency in seconds.",
    ["route", "method"],
)

# --- Async worker (Kafka consumer) ---
events_consumed_total = Counter(
    "events_consumed_total",
    "Total number of Kafka events consumed, by topic and result.",
    ["topic", "result"],
)

event_processing_duration_seconds = Histogram(
    "event_processing_duration_seconds",
    "Time spent processing a consumed Kafka event, in seconds.",
    ["topic"],
)

notifications_sent_total = Counter(
    "notifications_sent_total",
    "Total number of notifications 'sent'.",
)


def route_template(request) -> str:
    """Return the matched route's path template, or 'other' if unmatched (404)."""
    route = request.scope.get("route")
    # Starlette routes expose `.path_format` (e.g. "/notifications").
    return getattr(route, "path_format", None) or "other"

File: services/api-gateway/metrics.go
package main

import (
	"context"
	"log/slog"
	"net/http"
	"strconv"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
)

// RED metrics (Rate, Errors, Duration) exposed on /metrics. The `route` label is
// the matched route TEMPLATE — for the gateway that is the upstream PREFIX
// (e.g. "/api/cart"), never the raw path ("/api/cart/carts/5"), so cardinality
// stays bounded — this is what your SLO and Grafana RED dashboards consume.
var (
	httpRequestsTotal = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "http_requests_total",
			Help: "Total number of HTTP requests processed.",
		},
		[]string{"route", "method", "status"},
	)
	httpRequestDuration = promauto.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "http_request_duration_seconds",
			Help:    "HTTP request latency in seconds.",
			Buckets: prometheus.DefBuckets,
		},
		[]string{"route", "method"},
	)
)

// routeHolder lets a per-route wrapper communicate the matched route template
// back up to the metrics middleware (the handler runs inside the middleware, so
// we hand it a pointer to mutate rather than threading a new context back out).
type routeHolder struct{ name string }

// withRoute tags a handler with its route template for metrics labelling.
func withRoute(name string, fn http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		setRoute(r, name)
		fn(w, r)
	}
}

// setRoute records the route template for the current request so the metrics
// middleware labels it correctly. The proxy handler uses this to report the
// upstream prefix (e.g. "/api/cart") rather than the full path.
func setRoute(r *http.Request, name string) {
	if h, ok := r.Context().Value(routeKey).(*routeHolder); ok {
		h.name = name
	}
}

// statusRecorder captures the response status code so middleware can record it.
type statusRecorder struct {
	http.ResponseWriter
	status int
}

func (r *statusRecorder) WriteHeader(code int) {
	r.status = code
	r.ResponseWriter.WriteHeader(code)
}

// metricsMiddleware records RED metrics for every request, labelled by route.
func metricsMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		holder := &routeHolder{name: "other"}
		ctx := context.WithValue(r.Context(), routeKey, holder)

		start := time.Now()
		rec := &statusRecorder{ResponseWriter: w, status: http.StatusOK}
		next.ServeHTTP(rec, r.WithContext(ctx))

		httpRequestsTotal.WithLabelValues(holder.name, r.Method, strconv.Itoa(rec.status)).Inc()
		httpRequestDuration.WithLabelValues(holder.name, r.Method).Observe(time.Since(start).Seconds())
	})
}

// loggingMiddleware emits one structured log line per request with the trace id,
// and seeds the request context with that trace id for handler logs. It skips the
// high-frequency probe/metrics endpoints to avoid log spam.
func loggingMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		traceID := traceIDFrom(r)
		ctx := context.WithValue(r.Context(), traceIDKey, traceID)
		r = r.WithContext(ctx)

		switch r.URL.Path {
		case "/healthz", "/readyz", "/metrics":
			next.ServeHTTP(w, r)
			return
		}

		start := time.Now()
		rec := &statusRecorder{ResponseWriter: w, status: http.StatusOK}
		next.ServeHTTP(rec, r)
		slog.Info("request",
			"method", r.Method,
			"path", r.URL.Path,
			"status", rec.status,
			"duration_ms", time.Since(start).Milliseconds(),
			"trace_id", traceID,
		)
	})
}
